### PR TITLE
allow aliases to be handled by path-manager

### PR DESCRIPTION
**Pull request by AWare**
**Adding an alias when content is renamed**
This adds a `formAlias` parameter to the `POST` endpoint, it's not bound using play- but should be post update.

When the `formAlias` param is true, an update is sent to dynamo changing `type` to `alias` instead of deleting the path before the new path is added as canonical.

**Adding an alias**
`PUT` can now add aliases as well as canonical paths.

~**Deleting an alias**
A `DELETE` endpoint for path has been added, this checks that a path is of `alias` type and deletes it.~

Deleting aliases presents a risk of race condition if a path is moved in composer and then the alias is deleted and a new piece published on that path before the initial move reaches capi.

### CODE REVIEW

File: path-manager/app/controllers/PathManagerController.scala
Comment (twrichards) [inline path-manager/app/controllers/PathManagerController.scala]: do we need this `val`?
```suggestion
    request.body.asJson.map(_.as[PathRecord]) match {
```

File: path-manager/app/controllers/PathManagerController.scala
Comment (twrichards) [inline path-manager/app/controllers/PathManagerController.scala:60]: should this BadRequest message mention aliases (like the log line)

File: path-manager/app/controllers/PathManagerController.scala
Comment (twrichards) [inline path-manager/app/controllers/PathManagerController.scala]: should this be `fromAlias`?

File: path-manager/app/controllers/PathManagerController.scala
Comment (twrichards) [inline path-manager/app/controllers/PathManagerController.scala]: when you say it isn't binding? is that from postman, or from Composer UI (if so, query params need to be explicitly passed through in `composer-backend` as per https://github.com/guardian/flexible-content/pull/3555/commits/66a4cc9d33f9f8610c621957481f8cce986cc594 e.g. https://github.com/guardian/flexible-content/pull/3555/commits/b860c5b72a06d34b9937a67080002ad8a01e12ff#diff-f0283f6e89cc87b6244d311f848fb391R526)

File: path-manager/app/model/PathRecord.scala
Comment (twrichards) [inline path-manager/app/model/PathRecord.scala:20]: any reason for these changing order?

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: are we generating another 'short' url here for the alias? could it not re-use the short URL for the canonical, is that not effectively already an alias (just with a generated path)

**Review by twrichards — COMMENTED** [review]
Left some comments/questions.

It LGTM, and if it unblocks other work to integrate into the bigger picture (Composer etc.) then it gets my 👍  (FWIW).

I think we agreed separately that I'd have a look refactoring to transactions more generally and if I can make a breakthrough we can refactor this to be 'transactional' at some later point.

File: path-manager/app/controllers/PathManagerController.scala
Comment (twrichards) [reply path-manager/app/controllers/PathManagerController.scala]: that could be why it's not 'binding'?

Comment (AWare) [reply path-manager/app/controllers/PathManagerController.scala]: No, it's an instruction to form an alias from the previous canonical path.

File: path-manager/app/controllers/PathManagerController.scala
Comment (AWare) [reply path-manager/app/controllers/PathManagerController.scala]: I've been doing the HTTP by hand on this, and play wasn't binding the route like I expected. I'm going to revisit on your later play build. (Obviously before merging)

File: path-manager/app/model/PathRecord.scala
Comment (AWare) [reply path-manager/app/model/PathRecord.scala:20]: Yeah, to match the case class order- this specific ordering doesn't matter but doesn't match. (This is from an abandoned refactoring that destructured the case class- and this had me confused over field order)

File: path-manager/app/services/PathStore.scala
Comment (AWare) [reply path-manager/app/services/PathStore.scala]: Good spot, should have deleted after copy.

File: path-manager/app/controllers/PathManagerController.scala
Comment (AWare) [reply path-manager/app/controllers/PathManagerController.scala]: Nope.

File: path-manager/app/controllers/PathManagerController.scala
Comment (AWare) [reply path-manager/app/controllers/PathManagerController.scala:60]: Absolutely should.

**Comment by twrichards** [comment]
If you wanted to see if the upgrade helps with param binding issue, feel free to rebase onto my `upgrade-everything` branch (see https://github.com/guardian/path-manager/pull/32) and change the base branch of this PR to `upgrade-everything`...
image
... which will automatically get switched to `master` when my PR is merged (thanks to some recent magic from GitHub).

File: path-manager/app/services/PathStore.scala
Comment (jonathonherbert) [inline path-manager/app/services/PathStore.scala:126]: I don't think this method name could be improved, but after discussion with @AWare, the concept of updating a canonical with an alias involves some logic that bore explanation. Could we add a comment to explain this process?

**Review by jonathonherbert — COMMENTED** [review]
I'm not too familiar with the domain, but after going through with @AWare the logic here looks good.

I'd recommend adding unit tests given @twrichards' work on adding a harness.

File: path-manager/test/services/PathStoreTest.scala
Comment (twrichards) [inline path-manager/test/services/PathStoreTest.scala]:
```suggestion
        fail("first path wasn't actually created so can't test 'PathStore.registerAlias'")
```

File: path-manager/test/services/PathStoreTest.scala
Comment (twrichards) [inline path-manager/test/services/PathStoreTest.scala]:
```suggestion
        fail("first path wasn't actually created so can't test 'PathStore.updateCanonicalWithAlias'")
```

File: path-manager/app/controllers/PathManagerController.scala
Comment (twrichards) [reply path-manager/app/controllers/PathManagerController.scala]: I wonder if this is still too close to the `fromAlias` query param you mention in the description etc?

Comment (twrichards) [reply path-manager/app/controllers/PathManagerController.scala]: in-fact on closer inspection, I think maybe the PR description should read `formAlias`...

Comment (twrichards) [reply path-manager/app/controllers/PathManagerController.scala]: I also wonder if it ought to be `shouldFormAlias` just to imply the 'instruction' idea a little more explicitly

File: path-manager/conf/routes
Comment (twrichards) [inline path-manager/conf/routes]: given our discussions today (RE: draft vs live), do we want to remove this as an endpoint for the time being (but retain the functionality (if only for the tests)?

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: reading the functionality, does this not delete all aliases associated with the supplied path? 

If so could it be...
```suggestion
  def deleteAliasesAssociatedWithPath(path: String) = {
```

If not, should `pathItems` in the function body not be something like `maybePathItem` (with an exception or Either to handle the 'else scenario' of the path not being an alias or if no records match that path).

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: I'm not sure this does return the 'remaining' paths... it appears to map every entry in `pathItems` to its `identifier` regardless of whether it has been deleted

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]:
```suggestion
        logger.debug(s"deleting alias path [$path] for id [${item.identifier}]")
```

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: [minor] since we have an if/else block (and this is not the guard pattern of early returning) inverting the if/else and removing the `!` negation would make this easier to follow

Comment (twrichards) [reply path-manager/app/services/PathStore.scala]: alternatively we could add a new `isInvalid` method to `PathValidator` (a one liner that does the negation there)

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [reply path-manager/app/services/PathStore.scala]: also the logger message is ambiguous as its possible no paths were deleted at all (and probably not needed given the more explicit one inside the `map`)

File: path-manager/conf/routes
Comment (AWare) [reply path-manager/conf/routes]: Force anyone wanting to remove an alias to do it in dynamo? Probably a good way to make sure it's never misused.

File: path-manager/app/services/PathStore.scala
Comment (AWare) [reply path-manager/app/services/PathStore.scala]: whoops, good catch

Comment (AWare) [reply path-manager/app/services/PathStore.scala]: Fixed, but probably should just delete.

File: path-manager/app/services/PathStore.scala
Comment (AWare) [reply path-manager/app/services/PathStore.scala]: No, it just deletes that one alias. The `map` is to handle an `Option`. This isn't clear, so I tidied it a bit. We might just want to delete this method after our chat today though.

File: path-manager/conf/routes
Comment (twrichards) [inline path-manager/conf/routes]: see my comment https://github.com/guardian/path-manager/pull/33#discussion_r467952889

Comment (twrichards) [reply path-manager/conf/routes]: also i think the PR description needs updating as it refers to `fromAlias`

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: think this line can be dropped as there's similar logging 207 below
```suggestion
```

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: Now you've explained elsewhere this is Option rather than list/seq, I don't think this needs to be original suggestion...
```suggestion
  def deleteAlias(path: String) = {
```

Comment (twrichards) [reply path-manager/app/services/PathStore.scala]: that said, based on your https://github.com/guardian/path-manager/pull/33#discussion_r468021596 if we want deletions only to happen direct in dynamo (I guess by one of us), then maybe we remove this function entirely for now and revisit if/when we need a feature for deleting aliases (where we implement the 'safe delete' I mentioned IRL yesterday - by checking CAPI Live or something)

File: path-manager/app/services/PathStore.scala
Comment (twrichards) [inline path-manager/app/services/PathStore.scala]: if we are keeping this then singular is probably more accurate...
```suggestion
        logger.debug(s"successfully deleted alias path [$path] for id [${id}]")
```

File: path-manager/app/services/PathStore.scala
Comment (AWare) [reply path-manager/app/services/PathStore.scala]: yeah, deleted.

File: path-manager/app/controllers/PathManagerController.scala
Comment (AWare) [reply path-manager/app/controllers/PathManagerController.scala]: I think getting it wrong in the PR desc is a good sign that it needs to be better, changed to `shouldFormAlias`

**Review by twrichards — APPROVED** [review]

**Review by twrichards — APPROVED** [review]
Awesome work Alex.

**Pull request merged**
